docs/idm-concept.md

```diff
@@ -0,0 +1,206 @@
+
+1. Both an Application and an IDM Wallet running in the same browser, in the same equipment:
+	- Solution: Use an iframe and the [postMessage API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage)
+2. An Application running in a browser and an IDM Wallet running as a native OS application, in the same equipment
+	- Solution: The IDM Wallet exposes a local [WebSocket](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API) server on a pre-defined port known by the IDM Client
+3. Both the Application and IDM Wallet are native OS applications, running on the same equipment
+	- Solution: Use OS specific inter-process communication mechanisms, such as [Distributed Objects](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DistrObjects/Concepts/AboutDistributedObjects.html) for MacOSx and [MIDL](https://docs.microsoft.com/en-us/windows/desktop/midl/midl-start-page) for Windows
+4. An Application running on a laptop and an IDM Wallet running in a smartphone
+    - Solution: Use IPFS's pubsub to find and talk to each other
+
+Conceptually, the IDM Bridge is composed by two parts: the provider-side and the consumer-side. The provider-side is embedded in the IDM Wallet while the consumer-side is embedded in the IDM Client. Each one of the sides have a set of transports that they support to communicate.
+
+Moreover, the IDM Client has a discovery mechanism that finds the most appropriate IDM Wallet to talk to, starting by locating one closer to its own context that also supports one of its transports. For reference, there are various degrees of closeness, from closest to the furthest: exactly the same context (e.g.: within same browser), same machine, same network or different network. The discovery is likely to be more transparent and automatic if both sides are close to each other. For example, in scenario `2`, the IDM Client can try to initiate a WebSocket connection to `localhost:<predefined-port>` to check if there's a IDM Wallet running there. On the contrary, the further both sides are, the less automatic the process is and might require users to mediate the process by scanning QR-codes, inputting numbers, or other mechanisms.
+
+Messages are exchanged from the IDM Client to the IDM Wallet and vice-versa through the IDM Bridge. Those messages will be defined as part of the [IDM Bridge spec](./spec/idm-spec#idm-bridge) to ensure the interoperability between different implementations.
+
+There will be a reference IDM Bridge written in JavaScript that will solve scenario `1` by leveraging the [postMessage API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) to create a communication channel between IDM Clients and a IDM Wallet running on a pre-defined domain. Later on, that bridge may also use the [Broadcast Channel API](https://developer.mozilla.org/en-US/docs/Web/API/Broadcast_Channel_API) to support wallets running in different domains.

```

Comment:
From a quick read of the Broadcast Channel API, it seems to require `same-origin-policy`:

 ```The Broadcast Channel API allows simple communication between browsing contexts (that is windows, tabs, frames, or iframes) with the same origin (usually pages from the same site).```

Reply:
@daviddahl You are right, it seems I confused this with something else or maybe there's no way to do postMessage to all tabs and wait for a responsive. Will investigate.

Reply:
It seems there isn't, meaning that the IDM Client may prompt the user to provide a set of domains, defaulting to the reference IDM Wallet UI domain. If other wallets become relevant, they might be pre-added to the whitelisted domains.

Reply:
I've removed the reference of the Broadcast Channel api.

Reply:
Cool. I would love for that `Broadcast Channel API` to work that way, but would be a security issue for sure. Perhaps we can figure out a way to leverage `Web Notifications`?

Reply:
Perhaps WebRTC might be used for this scenario, will investigate.

Reply:
To no block this PR, I've created an issue for this topic: https://github.com/ipfs-shipyard/pm-idm/issues/40